discovery/zookeeper/zookeeper.go

```diff
@@ -166,7 +166,13 @@ func (d *Discovery) Run(ctx context.Context, ch chan<- []*targetgroup.Group) {
 			tc.Stop()
 		}
 		// Drain event channel in case the treecache leaks goroutines otherwise.
-		for range d.updates {
+	L:
+		for {
+			select {
+			case <-d.updates:
```

Comment:
Why is this channel not getting closed?

Reply:
It was closed in a previous version (treecache.go) but at that moment it has generate a panic because of multiple close on the same channel. The same updates channel created in Discovery constructor has beep passed to multiple ZookeeperTreeCaches objects. 
When an stop event has come one tree cache has close the updates channel but the same will try the others tree caches and they will generate panic. Anyway, it was an unhappy solution to close a channel outside the code that created it when the channel is distributed to multiple goroutines.

That solution is explained in PR #2669.

Reply:
This seems racy, I think a better approach would be for Stop to wait until things are stopped.